components/zcash_address/src/encoding.rs

```diff
@@ -0,0 +1,318 @@
+use std::{convert::TryInto, error::Error, fmt, str::FromStr};
+
+use bech32::{self, FromBase32, ToBase32, Variant};
+
+use crate::{kind::*, AddressKind, Network, ZcashAddress};
+
+/// An error while attempting to parse a string as a Zcash address.
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    /// The string is an invalid encoding.
+    InvalidEncoding,
+    /// The string might be an unknown Zcash address from the future.
+    MaybeZcash,
+    /// The string is not a Zcash address.
+    NotZcash,
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParseError::InvalidEncoding => write!(f, "Invalid encoding"),
+            ParseError::MaybeZcash => write!(
+                f,
+                "This might be a Zcash address from the future that we don't know about"
+            ),
+            ParseError::NotZcash => write!(f, "Not a Zcash address"),
+        }
+    }
+}
+
+impl Error for ParseError {}
+
+impl FromStr for ZcashAddress {
+    type Err = ParseError;
+
+    /// Attempts to parse the given string as a Zcash address.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Remove leading and trailing whitespace, to handle copy-paste errors.
+        let s = s.trim();
+
+        // Most Zcash addresses use Bech32, so try that first.
+        match bech32::decode(s) {
+            // Zcash addresses only use the original Bech32 variant, since the data
+            // corresponding to a particular HRP always has a fixed length.
+            Ok((_, _, Variant::Bech32m)) => return Err(ParseError::NotZcash),
+            Ok((hrp, data, Variant::Bech32)) => {
+                // If we reached this point, the encoding is supposed to be valid Bech32.
+                let data =
+                    Vec::<u8>::from_base32(&data).map_err(|_| ParseError::InvalidEncoding)?;
+
+                let net = match hrp.as_str() {
+                    sapling::MAINNET | orchard::MAINNET => Network::Main,
+                    sapling::TESTNET | orchard::TESTNET => Network::Test,
+                    sapling::REGTEST | orchard::REGTEST => Network::Regtest,
+                    _ => {
+                        // Use some heuristics to try and guess whether this might be a Zcash
+                        // address from the future:
+                        // - Zcash HRPs always start with a 'z'.
```

Comment:
It's quite possible that we may want to add transparent address formats that use bech32 encoding. I would suggest that we allow "t" followed by any single character as an HRP for transparent addresses.